File: sennaar-rs/src/registry/entity.rs
use std::collections::HashMap;

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::cpl::CExpr;
use crate::registry::{Metadata, Type};
use crate::Identifier;

pub enum Entities<'de> {
    Typedef(Typedef<'de>),
    FunctionTypedef(FunctionTypedef<'de>),
    Command(Command<'de>),
}

pub trait Entity<'de> : Sized + Eq + Ord + Serialize + Deserialize<'de> {
    fn entity_metadata(&self) -> &HashMap<String, Metadata>;
    fn entity_metadata_mut(&mut self) -> &mut HashMap<String, Metadata>;

    fn has_metadata(&self, key: &str) -> bool {
        self.entity_metadata().contains_key(key)
    }

    fn try_get_metadata(&self, key: &str) -> Option<&Metadata> {
        self.entity_metadata().get(key)
    }

    fn get_metadata(&self, key: &str) -> &Metadata {
        self.try_get_metadata(key).unwrap()
    }

    fn get_string_metadata(&self, key: &str) -> Option<&String> {
        self.try_get_metadata(key).and_then(|meta| {
            if let Metadata::String { value } = meta {
                Some(value)
            } else {
                panic!("expected string metadata for key '{}', found {:?}", key, meta);
            }
        })
    }

    fn get_kvs_metadata(&self, key: &str) -> Option<&HashMap<String, Metadata>> {
        self.try_get_metadata(key).and_then(|meta| {
            if let Metadata::KeyValues { kvs } = meta {
                Some(kvs)
            } else {
                panic!("expected key-values metadata for key '{}', found {:?}", key, meta);
            }
        })
    }

    fn put_metadata_kv(&mut self, key: impl ToString, value: Metadata) {
        self.entity_metadata_mut().insert(key.to_string(), value);
    }

    fn put_metadata(&mut self, key: impl AsRef<str> + ToString) {
        let metadata_mut = self.entity_metadata_mut();
        if !metadata_mut.contains_key(key.as_ref()) {
            metadata_mut.insert(key.to_string(), Metadata::None);
        }
    }

    fn put_metadata_string(&mut self, key: impl ToString, value: impl ToString) {
        self.put_metadata_kv(key, Metadata::String { value: value.to_string() });
    }

    fn put_metadata_kvs(&mut self, key: impl ToString, kvs: HashMap<String, Metadata>) {
        self.put_metadata_kv(key, Metadata::KeyValues { kvs });
    }
}

include!("../macross.rs");
include!("entity_macross.rs");

entity!{EntityBase,}

entity!{
    Typedef<'a>,
    target: Type<'a>,
}

ss_enum! {
    Bitwidth, Bit32, Bit64
}

entity!{
    Bitmask<'a>,
    bitwidth: Bitwidth,
    bitflags: Vec<Bitflag<'a>>
}

entity!{
    Bitflag<'a>,
    value: CExpr<'a>
}

entity!{
    Command<'a>,
    params: Vec<Param<'a>>,
    result: Type<'a>,
    success_codes: Vec<CExpr<'a>>,
    error_codes: Vec<CExpr<'a>>,
    alias_to: Option<Identifier>
}

impl<'a> Command<'a> {
    pub fn sanitize(&self) {
        for param in &self.params {
            param.sanitize();
        }
    }

    pub fn sanitize_fix(&mut self) {
        for param in &mut self.params {
            param.sanitize_fix();
        }
    }
}

entity!{
    Param<'a>,
    ty: Type<'a>,
    optional: bool,
    len: Option<CExpr<'a>>
}

impl<'a> Param<'a> {
    pub fn sanitize(&self) {
        if let Type::PointerType(ptr_type) = &self.ty {
            assert_eq!(ptr_type.nullable, self.optional);
        }
    }

    pub fn sanitize_fix(&mut self) {
        if let Type::PointerType(ptr_type) = &mut self.ty {
            ptr_type.nullable = self.optional;
        }
    }
}

entity!{
    Constant<'a>,
    ty: Type<'a>,
    expr: CExpr<'a>,
}

entity!{
    Enumeration<'a>,
    variants: Vec<EnumVariant<'a>>,
}

entity!{
    EnumVariant<'a>,
    value: CExpr<'a>
}

entity!{
    FunctionTypedef<'a>,
    params: Vec<Param<'a>>,
    result: Type<'a>,
    is_pointer: bool,
    is_native_api: bool
}

impl<'a> FunctionTypedef<'a> {
    pub fn sanitize(&self) {
        for param in &self.params {
            param.sanitize();
        }
    }

    pub fn sanitize_fix(&mut self) {
        for param in &mut self.params {
            param.sanitize_fix();
        }
    }
}

entity!{OpaqueTypedef,}

entity!{OpaqueHandleTypedef,}

entity!{
    Structure<'a>,
    members: Vec<Member<'a>>,
}

entity!{
    Member<'a>,
    ty: Type<'a>,
    bits: Option<usize>,
    init: Option<CExpr<'a>>,
    optional: bool,
    len: Option<CExpr<'a>>
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[derive(JsonSchema)]
pub struct Import {
    pub name: Identifier,
    pub version: Option<String>,
    pub depend: bool
}

impl PartialEq for Import {
    fn eq(&self, other: &Self) -> bool {
        self.name == other.name && self.version == other.version
    }
}

impl Eq for Import {}

impl PartialOrd for Import {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Import {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.name.cmp(&other.name)
    }
}

File: sennaar-kt/src/main/kotlin/registry/entity.kt
package club.doki7.sennaar.registry

import club.doki7.sennaar.Identifier
import club.doki7.sennaar.cpl.CExpr
import club.doki7.sennaar.interned
import kotlinx.serialization.Serializable

sealed class Entity {
    abstract var name: Identifier

    var metadata: MutableMap<String, Metadata> = mutableMapOf()
    var doc: MutableList<String> = mutableListOf()
    var platform: Platform? = null

    final override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Entity || this.javaClass != other.javaClass) return false
        return name == other.name
    }

    final override fun hashCode(): Int = name.hashCode()

    fun rename(newName: String) {
        name.rename(newName)
    }

    fun rename(renamer: (String) -> String) {
        name.rename(renamer(name.value))
    }
}

@Serializable
data class Typedef(
    override var name: Identifier,
    var target: Type
) : Entity() {
    constructor(
        name: String,
        target: Type
    ) : this(name.interned(), target)
}

@Serializable
enum class Bitwidth { Bit32, Bit64 }

@Serializable
data class Bitmask(
    override var name: Identifier,
    var bitwidth: Bitwidth,
    var bitflags: MutableList<Bitflag>
) : Entity() {
    constructor(
        name: String,
        bitwidth: Bitwidth,
        bitflags: MutableList<Bitflag>
    ) : this(name.interned(), bitwidth, bitflags)
}

@Serializable
data class Bitflag(
    override var name: Identifier,
    var value: CExpr
) : Entity() {
    constructor(
        name: String,
        value: CExpr
    ) : this(name.interned(), value)
}

@Serializable
data class Command(
    override var name: Identifier,
    var params: MutableList<Param>,
    var result: Type,
    var successCodes: MutableList<CExpr>,
    var errorCodes: MutableList<CExpr>,
    var aliasTo: Identifier?
) : Entity() {
    constructor(
        name: String,
        params: MutableList<Param>,
        result: Type,
        successCodes: MutableList<CExpr>,
        errorCodes: MutableList<CExpr>,
        aliasTo: Identifier? = null
    ) : this(name.interned(), params, result, successCodes, errorCodes, aliasTo)

    fun sanitize() {
        params.forEach { it.sanitize() }
    }

    fun sanitizeFix() {
        params.forEach { it.sanitizeFix() }
    }
}

@Serializable
data class Param(
    override var name: Identifier,
    var ty: Type,
    var optional: Boolean,
    var len: CExpr?
) : Entity() {
    constructor(
        name: String,
        ty: Type,
        optional: Boolean,
        len: CExpr?
    ) : this(name.interned(), ty, optional, len)

    fun sanitize() {
        val ptrType = ty as? PointerType
        if (ptrType != null && ptrType.nullable != optional) {
            error("for parameter $name of pointer type, pointer nullability (${ptrType.nullable}) does not match parameter optionality (${optional})")
        }
    }

    fun sanitizeFix() {
        val ptrType = ty as? PointerType
        if (ptrType != null && ptrType.nullable != optional) {
            ptrType.nullable = optional
        }
    }
}

@Serializable
data class Constant(
    override var name: Identifier,
    var ty: Type,
    var expr: CExpr
) : Entity() {
    constructor(
        name: String,
        ty: Type,
        expr: CExpr
    ) : this(name.interned(), ty, expr)
}

@Serializable
data class Enumeration(
    override var name: Identifier,
    var variants: MutableList<EnumVariant>
) : Entity() {
    constructor(
        name: String,
        variants: MutableList<EnumVariant>
    ) : this(name.interned(), variants)
}

@Serializable
data class EnumVariant(
    override var name: Identifier,
    var value: CExpr
) : Entity() {
    constructor(
        name: String,
        value: CExpr
    ) : this(name.interned(), value)
}

@Serializable
data class FunctionTypedef(
    override var name: Identifier,
    var params: MutableList<Param>,
    var result: Type,
    var isPointer: Boolean,
    var isNativeAPI: Boolean
) : Entity() {
    constructor(
        name: String,
        params: MutableList<Param>,
        result: Type,
        isPointer: Boolean,
        isNativeAPI: Boolean
    ) : this(name.interned(), params, result, isPointer, isNativeAPI)

    fun sanitize() {
        params.forEach { it.sanitize() }
    }

    fun sanitizeFix() {
        params.forEach { it.sanitizeFix() }
    }
}

@Serializable
data class OpaqueTypedef(override var name: Identifier) : Entity() {
    constructor(name: String) : this(name.interned())
}

@Serializable
data class OpaqueHandleTypedef(override var name: Identifier) : Entity() {
    constructor(name: String) : this(name.interned())
}

@Serializable
data class Structure(
    override var name: Identifier,
    var members: MutableList<Member>
) : Entity() {
    constructor(
        name: String,
        members: MutableList<Member>
    ) : this(name.interned(), members)
}

@Serializable
data class Member(
    override var name: Identifier,
    var ty: Type,
    var bits: Int?,
    var init: CExpr?,
    var optional: Boolean,
    var len: CExpr?
) : Entity() {
    constructor(
        name: String,
        ty: Type,
        bits: Int?,
        init: CExpr?,
        optional: Boolean,
        len: CExpr?
    ) : this(name.interned(), ty, bits, init, optional, len)
}

@Serializable
data class Import(
    var name: String,
    var version: String?,
    var depend: Boolean
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Import) return false
        return name == other.name && version == other.version && depend == other.depend
    }

    override fun hashCode(): Int {
        var result = name.hashCode()
        result = 31 * result + (version?.hashCode() ?: 0)
        result = 31 * result + depend.hashCode()
        return result
    }

    operator fun compareTo(other: Import): Int {
        return name.compareTo(other.name)
    }
}
